### PR TITLE
Cryptol what4 symbolic simulator

### PR DESCRIPTION
**Pull request by robdockins**
New symbolic evaluation backend based on What4.  Currently sort-of works for basic usage with a fixed solver.

Additional work required to allow configuring which solver to use, to support the "any" portfolio solver, and to support multi-sat queries.

Note: this is based on the `eval-refactor` branch, and will be a much smaller PR once #684 is merged into master.

### CODE REVIEW

File: src/Cryptol/REPL/Monad.hs [hunk 1]
```diff
@@ -793,20 +794,27 @@ checkInfLength val = case val of
 checkProver :: Checker
 checkProver val = case val of
   EnvString s
-    | s `notElem` SBV.proverNames ->
-      noWarns $ Just $ "Prover must be " ++ proverListString
-    | s `elem` ["offline", "any"] -> noWarns Nothing
-    | otherwise ->
+    | s `elem` ["offline", "any", "w4-offline"] -> noWarns Nothing
```
Comment (robdockins) [inline src/Cryptol/REPL/Monad.hs:797]: Add alternate names "sbv-offline" and "sbv-any"

File: src/Cryptol/Symbolic/What4.hs [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+  ]
+
+proverNames :: [String]
+proverNames = map fst proverConfigs
+
+lookupProver :: String -> IO (SolverAdapter st)
+lookupProver s =
+  case lookup s proverConfigs of
+   Just cfg -> pure cfg
+   Nothing  -> fail ("Invalid prover: " ++ s)
```
Comment (robdockins) [inline src/Cryptol/Symbolic/What4.hs:99]: Note to self: remove `fail` in favor of `panic` or throwing exceptions

File: src/Cryptol/Symbolic/What4.hs [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+  | VarFinSeq Int [VarShape sym]
+  | VarTuple [VarShape sym]
+  | VarRecord (Map.Map Ident (VarShape sym))
+
+freshVariable :: W4.IsSymExprBuilder sym => sym -> FinType -> IO (VarShape sym)
+freshVariable sym ty =
+  case ty of
+    FTBit         -> VarBit      <$> W4.freshConstant sym W4.emptySymbol W4.BaseBoolRepr
+    FTInteger     -> VarInteger  <$> W4.freshConstant sym W4.emptySymbol W4.BaseIntegerRepr
+    FTIntMod 0    -> VarInteger  <$> W4.freshConstant sym W4.emptySymbol W4.BaseIntegerRepr
```
Comment (robdockins) [inline src/Cryptol/Symbolic/What4.hs:357]: 0 modulus isn't allowed, panic here instead

**Comment by robdockins** [comment]
I think this is now ready to go

**Comment by atomb** [comment]
Have you looked at all at the polarities in blocking predicates I commented on above? It could well be correct, as-is, but I just wanted to be sure you'd taken a look and convinced yourself that it is.

**Comment by robdockins** [comment]
I'm not seeing a comment from you... what's the question about blocking predicates?

**Comment by atomb** [comment]
Oops. My review from a while back wasn't submitted. Now it is.

File: src/Cryptol/Symbolic/What4.hs [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+
+computeBlockingPred ::
+  sym ~ W4.ExprBuilder t CryptolState fm =>
+  sym ->
+  W4.GroundEvalFn t ->
+  [VarShape sym] ->
+  IO (W4.Pred sym)
+computeBlockingPred sym evalFn vs =
+  do ps <- mapM (varBlockingPred sym evalFn) vs
+     foldM (W4.orPred sym) (W4.falsePred sym) ps
```
[the remark this answers is not in the document]
Comment (robdockins) [reply src/Cryptol/Symbolic/What4.hs:302]: We're computing a blocking predicate for a single solution; so if `x` and `y` are variables and we found a solution where `x = a` and `y = b`, to get a different solution we need to say that at least one variable is different, so either `(not (x = a /\ y = b))`, or equivalently, `x != a \/ y != b`.

All the blocking predicates for previous solutions get asserted individually in the following query, which is the same as conjoining them together.

File: src/Cryptol/REPL/Monad.hs [hunk 5]
```diff
@@ -793,20 +794,27 @@ checkInfLength val = case val of
...
+    | s `elem` W4.proverNames ->
+      do -- TODO! Implement solver installation testing for What4
```
[the remark this answers is not in the document]
Comment (robdockins) [reply src/Cryptol/REPL/Monad.hs:799]: Yeah, probably.  There's this: https://github.com/GaloisInc/what4/issues/32 but we should probably have a Cryptol issue too.

File: CHANGES.md [hunk 6]
```diff
@@ -1,3 +1,19 @@
...
+  symbolic simulator is used when selecting solvers with the `w4`
+  prefix, such as `w4-z3`, `w4-cvc4`, `w4-yices`, etc.
+
+* More detailed information about the status of various symbols
+  in the output of the `:browse` command (issue #688).
+
+## Bug fixes
+
+* Closed issues #346, #444, #614, #617, #660, #662, #663, #664, #667, #670
+
```
[the remark this answers is not in the document]
Comment (robdockins) [reply CHANGES.md:16]: 👍

File: src/Cryptol/Symbolic/What4.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (atomb) [reply src/Cryptol/Symbolic/What4.hs:302]: Ah, right. This is multiple variables for a single solution, not multiple solutions. Got it.

File: src/Cryptol/REPL/Monad.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (robdockins) [reply src/Cryptol/REPL/Monad.hs:799]: CF #705

**Pull request merged**